### PR TITLE
api generator: support default values

### PR DESCRIPTION
**Pull request by nsams**
Usage: set a default value in the entitiy, it will be used as default in the input. Null for relations is also supported, set undefined as default value in entity.

### CODE REVIEW

File: demo/api/src/products/entities/product.entity.ts [hunk 1]
```diff
@@ -130,22 +130,23 @@ export class Product extends BaseEntity<Product, "id"> implements DocumentInterf
 
     @Property({ type: "json" })
     @Field(() => [ProductDiscounts])
-    discounts: ProductDiscounts[] = [];
+    discounts?: ProductDiscounts[] = [];
```
Comment (johnnyomair) [inline demo/api/src/products/entities/product.entity.ts:133]: Why do you need to make this field optional?

File: demo/api/src/products/entities/product.entity.ts [hunk 2]
```diff
@@ -130,22 +130,23 @@ export class Product extends BaseEntity<Product, "id"> implements DocumentInterf
...
+
     @Property({ type: "json" })
     @Field(() => [String])
-    articleNumbers: string[] = [];
+    articleNumbers?: string[] = [];
 
     @Property({ type: "json", nullable: true })
     @Field(() => ProductDimensions, { nullable: true })
-    dimensions?: ProductDimensions;
+    dimensions?: ProductDimensions = undefined;
```
Comment (johnnyomair) [inline demo/api/src/products/entities/product.entity.ts:141]: Why don't we use null? The field will be null in the database. So the correct definition would be `dimensions: ProductDimensions | null = null`. Or did we decide to ignore this inconsistency? I think we've already talked about this a view times...

File: packages/api/cms-api/src/generator/generate-crud-input.ts [hunk 3]
```diff
@@ -50,32 +56,49 @@ export async function generateCrudInput(
             //skip those (TODO find a non-magic solution?)
             continue;
         } else if (prop.enum) {
+            const initializer = morphTsProperty(prop.name, metadata).getInitializer()?.getText();
+            const defaultValue = prop.nullable && (initializer == "undefined" || initializer == "null") ? "null" : initializer;
```
Comment (johnnyomair) [inline packages/api/cms-api/src/generator/generate-crud-input.ts:60]: I see, we support setting `null` as default value as well 😁

File: packages/api/cms-api/src/generator/generate-crud-input.ts [hunk 4]
```diff
@@ -50,32 +56,49 @@ export async function generateCrudInput(
...
             decorators.push(`@Field(${prop.nullable ? "{ nullable: true }" : ""})`);
             type = "Date";
         } else if (prop.type === "BooleanType" || prop.type === "boolean") {
+            const initializer = morphTsProperty(prop.name, metadata).getInitializer()?.getText();
+            const defaultValue =
+                prop.nullable && (initializer == "undefined" || initializer == "null")
+                    ? "null"
+                    : initializer == "true" || initializer == "false"
+                    ? initializer
+                    : undefined;
```
Comment (johnnyomair) [inline packages/api/cms-api/src/generator/generate-crud-input.ts:98]: Why can't this be `const defaultValue = prop.nullable && (initializer == "undefined" || initializer == "null") ? "null" : initializer;` as in the other cases?

File: demo/api/src/products/entities/product.entity.ts (hunk 1 above)
Comment (nsams) [reply demo/api/src/products/entities/product.entity.ts:133]: I don't need to but I decided to do so:
- to have an example where defalutValues is tested
- (more importantly): to make the admin-generator PR work (as it doesn't support this field)

File: packages/api/cms-api/src/generator/generate-crud-input.ts (hunk 4 above)
Comment (nsams) [reply packages/api/cms-api/src/generator/generate-crud-input.ts:98]: My initial idea was to support only literals but I gave up on that for the other cases. Could be changed here for consistency.

Comment (johnnyomair) [reply packages/api/cms-api/src/generator/generate-crud-input.ts:98]: Yes, please change it 🙂

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**